File: Neetcode Questions/Trees/count-good-nodes-in-binary-tree.cpp

// https://leetcode.com/problems/count-good-nodes-in-binary-tree/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    bool check(vector<int>&arr){
        int n = arr.size();
        int X = arr[n-1];
        for(auto ele : arr){
            if(ele > X) return false;
        }
        return true;
    }
    
    void solve(TreeNode*root, vector<int>arr, int &ans){
        
        if(!root) return;
        
        arr.push_back(root->val);
        if(check(arr)) ans++;
        
        solve(root->left, arr, ans);
        solve(root->right, arr, ans);
        arr.pop_back();
        return;
    }
    
    int goodNodes(TreeNode* root) {
        vector<int> arr;
        int ans = 0;
        if(!root) return ans;
        solve(root, arr, ans);
        return ans;
    }
};

File: Neetcode-150-js/Trees/count-good-nodes-in-binary-tree.js

// https://leetcode.com/problems/count-good-nodes-in-binary-tree/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var goodNodes = function(root) {
    if(!root) return 0;
    let goodNodes = 1;
    const path = [];
    path.push(root.val);

    function solve(root){
        if(!root) return;

        path.push(root.val);

        // checking for whether the node pushed is good or not ?
        var check = () =>  {
            let val = path[path.length-1];
            let flag = true;
            for(let i = path.length-2; i >= 0; i--){
                if(path[i] > val){
                    flag = false;
                    break;
                }
            }
            if(flag) goodNodes++;
        }

        check();
        solve(root.left);
        solve(root.right);
        path.pop();
        return;
    }

    solve(root);
    return goodNodes-1;
};